astropy/table/table.py

```diff
@@ -801,58 +828,97 @@ def _init_from_list(self, data, names, dtype, n_cols, copy):
         def_names = _auto_names(n_cols)
 
         for col, name, def_name, dtype in zip(data, names, def_names, dtype):
-            # Structured ndarray gets viewed as a mixin unless already a valid
-            # mixin class
-            if (isinstance(col, np.ndarray) and len(col.dtype) > 1 and
-                    not self._add_as_mixin_column(col)):
-                col = col.view(NdarrayMixin)
-
-            if isinstance(col, Column):
-                # If col is a subclass of self.ColumnClass, then "upgrade" to ColumnClass,
-                # otherwise just use the original class.  The most common case is a
-                # table with masked=True and ColumnClass=MaskedColumn.  Then a Column
-                # gets upgraded to MaskedColumn, but the converse (pre-4.0) behavior
-                # of downgrading from MaskedColumn to Column (for non-masked table)
-                # does not happen.
-                if issubclass(self.ColumnClass, col.__class__):
-                    col_cls = self.ColumnClass
-                else:
-                    col_cls = col.__class__
-
-                col = col_cls(name=(name or col.info.name or def_name),
-                              data=col, dtype=dtype,
-                              copy=copy, copy_indices=self._init_indices)
-
-            elif self._add_as_mixin_column(col):
-                # Copy the mixin column attributes if they exist since the copy below
-                # may not get this attribute.
-                if copy:
-                    col = col_copy(col, copy_indices=self._init_indices)
-
-                col.info.name = name or col.info.name or def_name
-
-            elif isinstance(col, np.ma.MaskedArray):
-                # Require that col_cls be a subclass of MaskedColumn, remembering
-                # that ColumnClass could be a user-defined subclass (though more-likely
-                # could be MaskedColumn).
-                col_cls = (self.ColumnClass
-                           if issubclass(self.ColumnClass, self.MaskedColumn)
-                           else self.MaskedColumn)
-                col = col_cls(name=(name or def_name), data=col, dtype=dtype,
-                              copy=copy, copy_indices=self._init_indices)
-
-            elif isinstance(col, np.ndarray) or isiterable(col):
-                col = self.ColumnClass(name=(name or def_name), data=col, dtype=dtype,
-                                       copy=copy, copy_indices=self._init_indices)
-
-            else:
-                raise ValueError('Elements in list initialization must be '
-                                 'either Column or list-like')
+            col = self._convert_data_to_col(col, copy, def_name, dtype, name)
```

Comment:
Gheez, I'm bikeshedding already: again a feel-free-to-ignore suggestion, maybe `_init_col_from_data` (or `._init_column`?), to more closely match how the class has `_init_from_list`, `_init_indices`, etc.?

Reply:
Continuing to nitpick away, to me conceptually `def_name` belongs here, not in the method, i.e., I'd do here
```
if col.info.name is None:
    col.info.name = def_name
```
If you do think it makes more sense in the method, I'd prefer to spell it out as `default_name`.

Reply:
`_convert_data_to_col` was inspired by, and paired with, the existing `_convert_col_for_table`.  Also, the `_init_*` methods are more in-place acting on `self`, so I'm happy enough with the naming here.

Reply:
@mhvk - about `def_name`, I did think about that because I was unhappy in the same way as you.  

However, I think that having a `def_name` arg will end up being the cleanest because `_convert_data_to_col` is taking care of copying.  In particular you cannot set `col.info.name = def_name` at this level because that will be corrupting the input `col`.

So meet you half way and change the name.

Reply:
Both OK (really was OK before, but even better now).